File: python/lazyk.py
#!/usr/bin/env python3

import sys
import io
from enum import Enum
from typing import BinaryIO, TextIO

EType = Enum(
    "EType", [
        "App", "S", "K", "I", "Iota", "CFalse",
        "S1", "S2", "K1", "I1",
        "CNum", "CNum1", "Inc", "Num", "Read",
    ],
)

class Expr:
    def __init__(
        self,
        type: EType,
        arg1: type["Expr"] | int | None = None,
        arg2: type["Expr"] | None = None,
    ) -> None:
        self.type = type
        self.arg1 = arg1
        self.arg2 = arg2

    def __str__(self):
        match (self.type):
            case EType.App:
                return "(%s %s)" % (self.arg1, self.arg2)
            case EType.S1 | EType.K1 | EType.I1:
                return "(%s %s)" % (self.type.name, self.arg1)
            case EType.S2:
                return "(%s %s %s)" % (self.type.name, self.arg1, self.arg2)
            case EType.CNum | EType.Num:
                return "%s#%d" % (self.type.name, self.arg1)
            case EType.CNum1:
                return "(%s#%d %s)" % (self.type.name, self.arg1, self.arg2)
            case _:
                return self.type.name

    def apply(self, arg: "Expr") -> "Expr":
        return Expr(EType.App, self, arg)

    def drop_i1(self) -> "Expr":
        cur = self
        while cur.type == EType.I1:
            cur = cur.arg1
        return cur


class Parser:
    def __init__(self, src: str) -> None:
        self.__input = io.StringIO(src)
        self.__unread = []

    def nextc(self) -> str:
        if len(self.__unread) > 0:
            return self.__unread.pop()
        while True:
            c = self.__input.read(1)
            while c == "#":
                self.__input.readline()
                c = self.__input.read(1)
            if c == "" or not c.isspace():
                return c

    def unreadc(self, c):
        if c != "":
            self.__unread.append(c)

    def read_cc(self, closingchar: str) -> Expr:
        c = self.nextc()
        if c == closingchar:
            return LazyK.I
        self.unreadc(c)
        e = self.read_one(False)
        while True:
            c = self.nextc()
            if c == closingchar:
                return e
            else:
                self.unreadc(c)
                e = e.apply(self.read_one(False))

    def read_one(self, i_is_iota: bool) -> Expr:
        c = self.nextc()
        match c:
            case "`" | "*":
                return self.read_app(c)
            case "(":
                return self.read_cc(")")
            case "S" | "s":
                return LazyK.S
            case "K" | "k":
                return LazyK.K
            case "I":
                return LazyK.I
            case "i":
                return LazyK.IOTA if i_is_iota else LazyK.I
            case "0" | "1":
                return self.read_jot(c)
            case "":
                raise SyntaxError("unexpected EOF")
            case _:
                raise SyntaxError(f'unexpected char "{c}"')

    def read_app(self, c0: str) -> Expr:
        stack = []
        buf = []
        is_iota = c0 == "*"
        while True:
            c = self.nextc()
            match c:
                case "`" | "*":
                    stack.append((is_iota, buf))
                    buf = []
                    is_iota = c == "*"
                case _:
                    self.unreadc(c)
                    buf.append(self.read_one(is_iota))
            while len(buf) == 2:
                e = buf[0].apply(buf[1])
                if len(stack) == 0:
                    return e
                else:
                    (is_iota, buf) = stack.pop()
                    buf.append(e)

    def read_jot(self, c0: str) -> Expr:
        c = c0
        e = LazyK.I
        while c == "0" or c == "1":
            if c == "0":
                e = e.apply(LazyK.S).apply(LazyK.K)
            else:
                e = LazyK.S.apply(LazyK.K.apply(e))
            c = self.nextc()
        self.unreadc(c)
        return e

    @staticmethod
    def parse(src: str) -> Expr:
        return Parser(src).read_cc("")


class LazyK:
    S = Expr(EType.S)
    K = Expr(EType.K)
    I = Expr(EType.I)
    CTRUE = K
    CFALSE = Expr(EType.CFalse)
    IOTA = Expr(EType.Iota)
    INC = Expr(EType.Inc)
    NUM0 = Expr(EType.Num, 0)
    CNUMS = [Expr(EType.CNum, i) for i in range(257)]

    def cn_to_int(self, e: Expr) -> int:
        result = self.eval_expr(e.apply(LazyK.INC).apply(LazyK.NUM0))
        if result.type != EType.Num:
            raise ValueError(f"invalid output format: {result}")
        else:
            return result.arg1

    def eval_expr(self, e: Expr) -> Expr:
        stack = []
        cur = e
        while True:
            cur = cur.drop_i1()
            while cur.type == EType.App:
                stack.append(cur)
                cur = cur.arg1.drop_i1()
            if len(stack) == 0:
                return cur
            a = cur
            cur = stack.pop()
            cur.arg1 = a
            self.eval_primitive(cur)

    def eval_primitive(self, e: Expr):
        lhs: Expr = e.arg1
        rhs: Expr = e.arg2.drop_i1()
        e.arg1 = e.arg2 = None
        match lhs.type:
            case EType.I:
                e.type = EType.I1
                e.arg1 = rhs
            case EType.CFalse:
                e.type = EType.I
            case EType.K:
                e.type = EType.K1
                e.arg1 = rhs
            case EType.K1:
                e.type = EType.I1
                e.arg1 = lhs.arg1
            case EType.S:
                e.type = EType.S1
                e.arg1 = rhs
            case EType.S1:
                e.type = EType.S2
                e.arg1 = lhs.arg1
                e.arg2 = rhs
            case EType.S2:
                e.arg1 = lhs.arg1.apply(rhs)
                e.arg2 = lhs.arg2.apply(rhs)
            case EType.Iota:
                e.arg1 = rhs.apply(LazyK.S)
                e.arg2 = LazyK.K
            case EType.Read:
                ch = LazyK.CNUMS[self.readc()]
                readnext = Expr(EType.Read)
                e.arg1 = rhs.apply(ch)
                e.arg2 = readnext
                lhs.type = EType.S2
                lhs.arg1 = Expr(EType.S2, LazyK.I, Expr(EType.K1, ch))
                lhs.arg2 = Expr(EType.K1, readnext)
            case EType.CNum:
                e.type = EType.CNum1
                e.arg1 = lhs.arg1
                e.arg2 = rhs
            case EType.CNum1:
                if lhs.arg2 == LazyK.INC:
                    if rhs.type != EType.Num:
                        rhs = self.eval_expr(rhs)
                    if rhs.type == EType.Num:
                        e.type = EType.Num
                        e.arg1 = lhs.arg1 + rhs.arg1
                    else:
                        raise ValueError(f"invalid output format: {rhs}")
                else:
                    f = lhs.arg2
                    x = rhs
                    for i in range(lhs.arg1):
                        x = f.apply(x)
                    e.type = x.type
                    e.arg1 = x.arg1
                    e.arg2 = x.arg2
            case EType.Inc:
                if rhs.type != EType.Num:
                    rhs = self.eval_expr(rhs)
                if rhs.type == EType.Num:
                    e.type = EType.Num
                    e.arg1 = rhs.arg1 + 1
                else:
                    raise ValueError(
                        f"invalid output format (attempted to apply inc to a non-number: {rhs})"
                    )
            case EType.Num:
                raise ValueError(f"invalid output format ({lhs} is not applicable)")
            case _:
                raise RuntimeError(f"unexpected state: lhs={lhs} rhs={rhs}")

    def readc(self) -> int:
        self.output.flush()
        b = self.input.read(1)
        if len(b) == 0:
            return 256
        c = b[0]
        if c == 13 and self.textmode:
            return self.readc()
        else:
            return c

    def writec(self, c: int):
        self.output.write(bytes([c]))
        if c == 10:
            self.output.flush()

    def print_list(self, lst: Expr) -> int:
        while True:
            c = self.cn_to_int(lst.apply(LazyK.CTRUE))
            if c >= 256:
                return (c - 256)
            self.writec(c)
            lst = lst.apply(LazyK.CFALSE)

    def __init__(
        self, input: BinaryIO, output: BinaryIO, textmode: bool = False
    ) -> None:
        self.input = input
        self.output = output
        self.textmode = textmode

    def run(self, src: str) -> int:
        prog = Parser.parse(src)
        return self.print_list(prog.apply(Expr(EType.Read)))

    def run_file(self, srcfile: str) -> int:
        with open(srcfile) as fp:
            return self.run(fp.read())


def __showhelp():
    print("Usage: python lazyk.py [-t] srcfile")
    print("Option")
    print("   -t : text mode (read crlf as lf)")
    sys.exit(1)


def __main(args):
    textmode = False
    srcfile = ""
    for a in args:
        if a == "-t":
            textmode = True
        else:
            srcfile = a
            break
    if srcfile == "":
        __showhelp()
    m = LazyK(sys.stdin.buffer, sys.stdout.buffer, textmode)
    try:
        if srcfile == "-":
            result = m.run(sys.stdin.read(-1))
        else:
            result = m.run_file(srcfile)
        return result
    except Exception as e:
        print(f"ERROR: {e}")
        return 1


if __name__ == "__main__":
    sys.exit(__main(sys.argv[1:]))

File: ruby/lazyk.rb
#!/usr/bin/env ruby

require "stringio"

class LazyK

  class Expr
    attr_accessor :type, :arg1, :arg2

    def initialize(t, a1 = nil, a2 = nil)
      @type = t
      @arg1 = a1
      @arg2 = a2
    end

    S = Expr.new(:s)
    K = Expr.new(:k)
    I = Expr.new(:i)
    IOTA = Expr.new(:iota)
    TRUE = K
    FALSE = Expr.new(:false)
    INC = Expr.new(:inc)
    NUM0 = Expr.new(:num, 0)
    CNUMS = (0..256).map{|i| Expr.new(:cn, i) }

    def drop_i1
      cur = self
      while cur.type == :i1
        cur = cur.arg1
      end
      cur
    end

    def apply(arg)
      Expr.new(:app, self, arg)
    end

    def eval()
      cur = self
      stack = []
      while true
        cur = cur.drop_i1
        while cur.type == :app
          stack << cur
          cur = cur.arg1.drop_i1
        end
        if stack.empty?
          return cur
        end
        a = cur
        cur = stack.pop
        cur.arg1 = a
        cur.eval_primitive
      end
    end

    def eval_primitive()
      lhs = @arg1
      rhs = @arg2.drop_i1
      @arg1 = @arg2 = nil
      case lhs.type
      when :false
        @type = :i
      when :i
        @type = :i1
        @arg1 = rhs
      when :k
        @type = :k1
        @arg1 = rhs
      when :k1
        @type = :i1
        @arg1 = lhs.arg1
      when :s
        @type = :s1
        @arg1 = rhs
      when :s1
        @type = :s2
        @arg1 = lhs.arg1
        @arg2 = rhs
      when :s2
        @arg1 = lhs.arg1.apply(rhs)
        @arg2 = lhs.arg2.apply(rhs)
      when :iota
        @arg1 = rhs.apply(Expr::S)
        @arg2 = Expr::K
      when :read
        reader = lhs.arg1
        cn = CNUMS[reader[]]
        readnext = Expr.new(:read, reader)
        @arg1 = rhs.apply(cn)
        @arg2 = readnext
        # S(SI(K c))(K r) = \b. b c r = cons c r
        lhs.type = :s2
        lhs.arg1 = Expr.new(:s2, Expr::I, Expr.new(:k1, cn))
        lhs.arg2 = Expr.new(:k1, readnext)
      when :cn
        @type = :cn1
        @arg1 = lhs.arg1
        @arg2 = rhs
      when :cn1
        if lhs.arg2.type == :inc
          rhs = rhs.eval if rhs.type != :num
          if rhs.type == :num
            @type = :num
            @arg1 = lhs.arg1 + rhs.arg1
          else
            raise "invalid output format (attempted to apply inc to a non-number): (#{rhs})"
          end
        else
          f = lhs.arg2
          x = rhs
          lhs.arg1.times{ x = f.apply(x) }
          @type = x.type
          @arg1 = x.arg1
          @arg2 = x.arg2
        end
      when :inc
        rhs = rhs.eval if rhs.type != :num
        if rhs.type == :num
          @type = :num
          @arg1 = rhs.arg1 + 1
        else
          raise "invalid output format (attempted to apply inc to a non-number): (#{rhs})"
        end
      when :num
        raise "invalid output format (attempted to apply a number): #{lhs} #{rhs}"
      else
        raise ScriptError.new("unexpected state: lhs=#{lhs} rhs=#{rhs}")
      end
    end

    def to_i
      e = self.apply(Expr::INC).apply(Expr::NUM0).eval
      if e.type == :num
        e.arg1
      else
        raise "invalid output format (result was not a number): #{e}"
      end
    end

    def to_s
      case @type
      when :app then "(#{@arg1.to_s} #{@arg2.to_s})"
      when :i1, :s1, :k1 then "(#{@type} #{@arg1.to_s})"
      when :s2 then "(S #{@arg1.to_s} #{@arg2.to_s})"
      when :cn then "CN#{@arg1}"
      when :cn1 then "(CN#{@arg1} #{@arg2.to_s})"
      when :num then @arg1.to_s
      when :iota then "Iota"
      else @type.to_s.upcase
      end
    end
  end

  class Parser
    def initialize(src)
      @input = StringIO.new(src)
    end

    def nextchar()
      begin
        c = @input.readchar
        while c == "#"
          @input.readline
          c = @input.readchar
        end
      rescue EOFError
        return nil
      end while c.match(/\s/)
      c
    end

    def read_cc(closingchar = nil)
      c = nextchar
      if (c == closingchar)
        return Expr::I
      end
      @input.ungetc(c)
      e = read_one(false)
      while (c = nextchar) != closingchar
        @input.ungetc(c)
        e = e.apply(read_one(false))
      end
      e
    end

    def read_one(i_is_iota)
      c = nextchar
      case c
      when "`", "*"
        read_app(c)
      when "("
        read_cc(")")
      when "S", "s"
        Expr::S
      when "K", "k"
        Expr::K
      when "I"
        Expr::I
      when "i"
        i_is_iota ? Expr::IOTA : Expr::I
      when "0", "1"
        read_jot(c)
      when nil
        raise SyntaxError.new("unexpected EOF")
      else
        raise SyntaxError.new("unexpected char: '${c}'")
      end
    end

    def read_app(c0)
      stack = []
      buf = []
      in_iota = c0 == "*"
      while true
        c = nextchar
        if c == "`" || c == "*"
          stack << [in_iota, buf]
          buf = []
          in_iota = c == "*"
        else
          @input.ungetc(c)
          buf << read_one(in_iota)
        end
        while buf.length == 2
          e = buf[0].apply(buf[1])
          if stack.empty?
            return e
          end
          in_iota, buf = stack.pop
          buf << e
        end
      end
    end

    def read_jot(c0)
      c = c0
      e = Expr::I
      while true
        if c == "0"
          e = e.apply(Expr::S).apply(Expr::K)
        elsif c == "1"
          e = Expr::S.apply(Expr::K.apply(e))
        else
          break
        end
        c = nextchar
      end
      @input.ungetc(c)
      e
    end

    def self.parse(src)
      Parser.new(src).read_cc
    end
  end

  def initialize(stdin = $stdin, stdout = $stdout, textmode = false)
    @stdout = stdout
    @stdout.binmode
    @stdin = stdin
    @textmode = textmode
  end

  def readc()
    begin
      @stdout.flush
      b = @stdin.readbyte
      if b == 13 && @textmode
        return readc
      else
        return b
      end
    rescue EOFError
      return 256
    end
  end

  def writec(c)
    @stdout.write([c].pack("C"))
    @stdout.flush if c == 10
  end

  def print_list(lst)
    car = ->(e) { e.apply(Expr::K).eval }
    cdr = ->(e) { e.apply(Expr::FALSE).eval }
    while true
      c = car[lst].to_i
      if c >= 256
        @stdout.flush
        return c-256
      else
        writec(c)
      end
      lst = cdr[lst]
    end
  end

  def run(src)
    prog = Parser.parse(src)
    input = Expr.new(:read, method(:readc))
    print_list(prog.apply(input))
  end

  def run_file(filepath)
    File.open(filepath, "r"){|f| run(f.read) }
  end
  public :run, :run_file
end

def show_help_and_exit
  puts "Usage: ruby [-t] lazyk.rb prog.rb"
  puts "Options"
  puts "  -t : text mode (read crlf as lf)"
  exit(1)
end

if __FILE__ == $0
  if ARGV.length == 0
    textmode = false
    if $stdin.tty?
      show_help_and_exit
    else
      exit LazyK.new(StringIO.new(""), textmode).run($stdin.read)
    end
  else
    if ARGV[0] == "-h" || ARGV[0] == "--help"
      show_help_and_exit
    else
      if ARGV[0] == "-t"
        textmode = true
        ARGV.shift
      end
      exit LazyK.new($stdin, $stdout, textmode).run_file(ARGV[0])
    end
  end
end
